### DISCUSSION TITLE
primaryjoin and delete

### DISCUSSION THREAD

**Discussion by ljiatu** [sqlalchemy/sqlalchemy#5983] [Usage Questions] [answered]
Hi there,

We are observing something with primaryjoin, raise lazy load and delete and not sure if this is the expected behavior (it may very well is, but I struggled to find where it is documented).

We have the following two tables:

```
class User(Base, TimestampColumns):
    email = Column(
        String, unique=True, index=True
    )
    
    <...unrelated fields omitted>
    
    email_confirmation_keys = relationship(
        "EmailConfirmationKey",
        backref=backref("user", lazy="raise"),
        cascade="all, delete-orphan",
        passive_deletes=True,
        lazy="raise",
        primaryjoin="EmailConfirmationKey.email == User.email",
    )

class EmailConfirmationKey(Base, TimestampColumns):
    email = Column(
        String,
        ForeignKey("users.email", onupdate="CASCADE", ondelete="CASCADE"),
        nullable=False,
        index=True,
        unique=True,
    )
```

And we have a simple function that deletes an email confirmation key:
```
def delete_key(session: Session, key: str) -> None:
    eck = (
        session.query(EmailConfirmationKey).filter(EmailConfirmationKey.id == key).one()
    )
    session.delete(eck)
    session.commit()
```

However, it errors on `commit`: `sqlalchemy.exc.InvalidRequestError: 'EmailConfirmationKey.user' is not available due to lazy='raise'`. If I use the `users.id` column in the `email_confirmation_keys` as the foreign key column, then the problem goes away. So it seems like `primaryjoin` on a non-primary key column makes it load the relationship on delete. Is this expected?

Thanks!

**Comment by zzzeek** [comment, marked as answer]
hey there -

please send complete stack traces and sample scripts for these.  I've reconstructed it as below:
```

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import relationship, backref
from sqlalchemy.orm import Session
Base = declarative_base()

class User(Base):
    __tablename__ = 'users'

    id = Column(Integer, primary_key=True)
    email = Column(
        String, unique=True, index=True
    )

    email_confirmation_keys = relationship(
        "EmailConfirmationKey",
        backref=backref("user", lazy="raise"),
        cascade="all, delete-orphan",
        passive_deletes=True,
        lazy="raise",
        primaryjoin="EmailConfirmationKey.email == User.email",
    )

class EmailConfirmationKey(Base):
    __tablename__ = 'email'
    id = Column(Integer, primary_key=True)
    email = Column(
        String,
        ForeignKey("users.email", onupdate="CASCADE", ondelete="CASCADE"),
        nullable=False,
        index=True,
        unique=True,
    )


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

u1 = User(email="x", email_confirmation_keys=[EmailConfirmationKey()])
s.add(u1)

s.commit()


e1 = s.query(EmailConfirmationKey).first()
s.delete(e1)
s.flush()



```

you can prevent the ORM from caring about EmailConfirmationKey.user by placing passive_deletes=True on the relationship:

```
    email_confirmation_keys = relationship(
        "EmailConfirmationKey",
        backref=backref("user", lazy="raise", passive_deletes=True),
        cascade="all, delete-orphan",
        passive_deletes=True,
        lazy="raise",
        primaryjoin="EmailConfirmationKey.email == User.email",
    )

```

the reason it's attempting to load EmailConfirmationKey.user is because the primaryjoin condition is a "non-simple" primaryjoin;  it iinvolves a custom condition to a non-primary key column.  when a many to one nonetheless has a complex join, SQLAlchemy assumes it might have anything inside of it which it might need to attend towards when it is deleting the row.     I just tried disabling this call for all many-to-one with isdelete and some tests do fail without, tests which involve cyclical foreign keys pointing in both directions to each other, which is the kind of thing this is meant for.    Due to the introudction of lazy="raise", tuning the delete process to be more sensitive to edge cases like this has become more important.  I will add this case as a bug but it's lower proirity for the moment.

**Reply by zzzeek** [reply]
#5984

**Reply by ljiatu** [reply]
Thanks for the fast response as always! I will send a more complete script next time. Thanks for the reminder.

So I do have `passive_deletes=True` in my original code. Do you mean it shouldn't load the `user` if that flag is enabled (so this is a bug) or it would still load the `user` even if that flag is `True`?

**Reply by zzzeek** [reply]
I added it to the backref which is where it was needed for this particular load.   a "backref" is another "relationship".

in this case, the passive_deletes=True (on the backref) for me prevents it from doing the load, whch is what it's supposed to do here.

what I'm doing in #5984 is disabling the entire "raiseload" thing from happening in the flush process, that is, if it needs to load something while it's trying to flush, it will load it rather than raise an error.   "raiseload" was more about end-user code directly triggering attributes.   as you are a user of "lazy='raise'" does this fit with your idea of it?

**Reply by ljiatu** [reply]
Ah got it. I overlooked the `passsive_deletes=True` in `backref`. Thanks for pointing it out!

Re: #5984, I think your approach makes sense, but it might be confusing to some people since it seems to contradict what `lazy="raise"` is supposed to prevent, e.g. I was under the assumption that `raiseload` would prevent all sorts of lazy load no matter it's the end-user code explicitly requesting the attributes or some library code, until you clarified above. So I think some documentation would be needed if this new feature gets shipped.

Also, I think some documentation around non-trivial `primaryjoin` triggering relationship attributes load during the delete process would be great!

**Reply by zzzeek** [reply]
it's very difficult to document things like that since we dont have an overarching document of every single thing the unit of work does, if we did it would be very complicated, and additionally it changes when we encounter bugs like this one.  so it's really a moving target and that's why having raiseload be a part of it makes raiseload itself less useful.

also, it shouldnt raelly have to load the relationship here and i will probably figure out a way to get it to skip this one, it's only when there are "post update" options that it seems to really be needed.

the docs will clarify that raiseload is only for end-user relationships.  there's precedent already for this as there are  already some history-related spots that skip raiseload.

**Reply by ljiatu** [reply]
Yep, agree that it shouldn't need to load the attribute here. I think what you said makes sense about documentation. Thanks for looking into it!

**Reply by zzzeek** [reply]
OK, the load that it's doing is necessary if you didn't say passive_deletes=True because it needs to know about what is happening to that object, such as if this object has to be deleted after that one, as well as some other cases.   maybe it could be smarter to know ahead of time that this mapping wouldn't have further relationships to be deleted but right now the system doesn't have logic to think that far ahead for this kind of thing.   I might be able to optimize very simple cases a little bit when we can tell there's no other relationships to affect anything.

**Reply by ljiatu** [reply]
So we added `passive_deletes=True`, and it's working as expected, but SQLAlchemy is giving this warning:
```
/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/relationships.py:2273: SAWarning: On EmailConfirmationKey.user, 'passive_deletes' is normally configured on one-to-many, one-to-one, many-to-many relationships only.
```
I'm assuming we can just ignore it?

**Reply by zzzeek** [reply]
you can ignore that yes.     

in 1.4 I just took the "raise" out of the equation for things that happen in unit of work.   For skipping the load altogether, I'm sure this is doable but it was proving to be a very fickle set of conditions and I think once "raise" isn't affected people won't care much.

also I wonder how useful that warning is here.    it's likely people put the warning on the wrong side of relationships and were confused and that's why I added it.

**Reply by zzzeek** [reply]
this warning will be removed in the next 1.3 release https://gerrit.sqlalchemy.org/q/I07adb726633f73f26296add4ec70376d5dcf8817

**Reply by ljiatu** [reply]
Great, thank you!
